File: util/binomial_coefficient.cc
#include <iostream>
using namespace std;
const long long BIGINT = (long long)1e9 + 7;

long long pow(long long x, int n)
{
    long long res = 1LL;
    int acc = 1;
    for (int i = 0; acc <= n; ++i) {
        if ((n >> i) & 1) {
            res = (res * x) % BIGINT;
        }
        x = (x * x) % BIGINT;
        acc *= 2;
    }
    return res;
}

long long modinv(long long a, int p)
{
    return pow(a, p - 2);
}

long long binomial_coeff(int n, int r)
{
    if (n < r)
        return 0LL;
    if (2 * r > n)
        return binomial_coeff(n, n - r);

    long long numer = 1LL;
    long long denom = 1LL;
    int rr = r;
    for (int i = 0; i < rr; ++i) {
        numer = (numer * n) % BIGINT;
        denom = (denom * r) % BIGINT;
        n--;
        r--;
    }
    cerr << "  numer = " << numer << endl;
    cerr << "  denom = " << denom << endl;
    return (numer * modinv(denom, BIGINT)) % BIGINT;
}

File: util/rust/binomial_coefficient.rs
const BIGINT: i64 = 1e9 as i64 + 7;


fn length_in_base2(x: i64) -> i64 {
    let mut x = x;
    let mut ans = 0;
    while x > 0 {
        ans += 1;
        x >>= 1;
    }
    ans
}

fn powmod(x: i64, y: i64, z: i64) -> i64 {
    let mut ans = 1_i64;
    let mut px = x;
    let y_size = length_in_base2(y);
    for i in 0..y_size {
        if (y >> i) & 1 == 1 {
            ans = (ans * px) % z;
        }
        px = (px * px) % z;
    }
    ans
}

fn modinv(a: i64, p: i64) -> i64 {
    powmod(a, p - 2, p)
}

fn binomial_coeff(n: i64, r: i64, z:i64) -> i64 {
    if n < r {
        return 0;
    }
    if 2 * r > n {
        return binomial_coeff(n, n - r, z);
    }

    let mut numer = 1;
    let mut denom = 1;
    let mut n = n;
    let mut r = r;
    for _ in 0..r {
        numer = (numer * n) % z;
        denom = (denom * r) % z;
        n -= 1;
        r -= 1;
    }
    let ans = (numer * modinv(denom, z)) % z;
    ans
}

fn main() {
    assert_eq!(binomial_coeff(5, 2, BIGINT), 10);
    assert_eq!(binomial_coeff(64, 32, BIGINT), 114221638);
    assert_eq!(binomial_coeff(100, 20, BIGINT), 926116413);
}
